server/api/cvrs.py

```diff
@@ -408,6 +421,274 @@ def parse_cvr_rows() -> Iterable[CvrBallot]:
+        }
+
+        return {
+            contest_name: dict(
+                # Until we know how vote-for-n contests are serialized in the
+                # ES&S CVR, we assume vote-for-1
+                votes_allowed=1,
+                choices={
+                    choice: dict(
+                        column=contest_choice_columns[(contest_name, choice)],
+                        num_votes=0,  # Will be counted while parsing rows
+                    )
+                    for choice in sorted(choices)
+                },
+                total_ballots_cast=0,  # Will be counted while parsing rows
+            )
+            for contest_name, choices in contest_choices.items()
+            if len(choices) > 0
+        }
+
+    def parse_interpretations(
+        cvr_csv: CSVIterator, contests_metadata: CVR_CONTESTS_METADATA
+    ) -> Iterable[Tuple[str, str]]:  # (CVR number, interpretations)
+        # pylint: disable=stop-iteration-return
+        headers = next(cvr_csv)
+        header_indices = get_header_indices(headers)
+
+        max_interpretation_column = max(
+            choice_metadata["column"]
+            for contest_metadata in contests_metadata.values()
+            for choice_metadata in contest_metadata["choices"].values()
+        )
+
+        def parse_row_interpretations(row: List[str], cvr_number: int,) -> str:
+            interpretations = ["" for _ in range(max_interpretation_column + 1)]
+            for contest_name, contest_metadata in contests_metadata.items():
+                recorded_choice = column_value(
+                    row, contest_name, cvr_number, header_indices, required=False
+                )
+                if recorded_choice:
+                    for choice_name, choice_metadata in contest_metadata[
+                        "choices"
+                    ].items():
+                        if choice_name == recorded_choice:
+                            interpretations[choice_metadata["column"]] = "1"
+                        else:
+                            interpretations[choice_metadata["column"]] = "0"
+
+            return ",".join(interpretations)
+
+        for row_index, row in enumerate(cvr_csv):
+            try:
+                cvr_number = column_value(
+                    row, "Cast Vote Record", row_index + 1, header_indices
+                )
+                yield (cvr_number, parse_row_interpretations(row, cvr_number))
+            except UserError as error:
+                raise UserError(f"{cvr_file_name}: {error}") from error
+
+    def parse_and_concat_ballots_files(
+        ballots_files: Dict[str, TextIO]
+    ) -> Iterable[CvrBallot]:
+        for name, ballots_file in ballots_files.items():
+            try:
+                validate_comma_delimited(ballots_file)
+                ballots_csv = csv.reader(ballots_file, delimiter=",")
+                yield from parse_ballots_csv(ballots_csv)
+                ballots_file.close()
```

Comment:
It's a bit odd that `parse_and_concat_ballots_files` is responsible for closing the files. It looks like the opening happens in `unzip_files`? I'm unsure how to make this clearer, but I first noticed this because I wondered what would happen to the files given an exception. Maybe we don't care as we're not handling it and it'll just bubble up.

Reply:
Yeah I'm not sure what a clear pattern for this would be. The only Python idiom I know is to use context managers for opening/closing files:

```python
with open(path) as file:
  ... do stuff ...
```

I started out trying to use this pattern, but if you want to `return` the file inside the `with`, it doesn't work, because then the file gets closed when the function returns. 

It's possible to make `unzip_files` into a context manager so that it can be used in a `with` block, but then everything has to be indented, which I think is a major knock on readability.

I guess one pattern could be:

```python
def do_all_the_things(zip_file):
  ... do stuff ...

with unzip_files(path) as zip_file:
  do_all_the_things(zip_file)
```

Do you think that would be more readable? Again, adds a level of indentation to make the helper function.

As far as what happens if you don't close files, I haven't been able to find evidence that there are any real consequences. My cursory research seemed to indicate that they just get closed when they are garbage collected. But might be good to put the `close` calls into a `finally` block or something.

Reply:
I don't feel strongly about it since this is all contained within the parent function.

Reply:
I decided to at least move the `close` call to a `finally` block. And also realized I never closed `cvr_file`, so added that. 4f77bf93